labs/b4.md

```diff
@@ -0,0 +1,167 @@
+
+## So how do I compile?
+
+Compiling software on Linux is a mixed bag. Sometimes, all the dependencies are installed on your computer and there is no fiddling around. Sometimes, the dependencies don’t even exist pre-compiled for your distribution so you have to compile those yourself in order to compile what’s at hand. But all of these steps all have one component in common: the makefile.
+
+In most source tarballs, there is usually a makefile that contains a set of directives to compile a project. This is because in a project, there is usually multiple files that need to be compiled by the compiler. And even then, there are multiple compilers and compiler settings that can be used.
+
+Many projects that have to be compiled are usually in C and/or C++. On most Linux distributions, there are usually three compiler “options”. There is the GNU Toolchain which uses GCC, LLVM which uses clang, and Intel’s proprietary toolkit. Both GCC and clang are open-source and free. However, clang was created with a permissive license in response to GCC’s more viral copyleft license. Although clang was made with the intention of being completely compatible with GCC, there are some quirks and most open-source software still target GCC.
+
+To compile most software that you have the dependencies to, run:
+
+`make`
+
+This is will usually choose the correct compiler and compile the whole project. Usually when the compilation is done, the executables are in `/bin` or `/src`. 
+
+GCC and clang also have compiler flags that allow certain features to be enabled. Usually the flags that actually matter are optimization flags. Depending on what you want to optimize for, either space or memory or speed, there is a flag for it.
+
+Now, we will make a very simple application in C that prints “Hello Penguin!” named `hellopenguin`. Run:
+
+`nano hellopenguin.c`
+
+to create a file named `hellopenguin.c`, and type in the following:
+
+```
+#include <stdio.h>
+int main()
+{
+   printf("Hello Penguin!");
+   return 0;
+}
+```
+
+Now save and exit. 
+
+We will now compile the source file that you have just written:
+
+`gcc hellopenguin.c -o hellopenguin`
+
+What this does is to take in a source file `hellopenguin.c` and compile it to an executable named `hellopenguin` with the `-o` output flag.
+
+# How do I package stuff?
+
+Packaging manually for Debian can be very hard and frustrating, especially for first timers. That’s why for this class, we’ll be using a really cool Ruby package called `fpm` which simplifies the task of packaging a lot. Please note that if you were going to package for Debian’s own repositories, this is NOT the correct/formal way to go about packaging. It is merely a great way to backport or package your own applications extremely quickly. 
```

Comment:
I'd try to avoid, in generaly, saying that these things are "very hard and frustrating", over something less severe like "Packacking manually can be frustrating for first-timers. That's why we'll be using..."